lib/eval.ml

```diff
@@ -881,32 +987,47 @@ module MakeInterpreter (T : Target) = struct
     let args' = List.map ~f:(fun arg -> Some arg) args in
     let (env',st',s,v) = let open Declaration in match snd decl with
       | Control typ_decl ->
-        let (env',st',s) = copyin ctrl env st typ_decl.constructor_params args' in
-        let state = env' |> EvalEnv.get_val_firstlevel |> List.rev in
-        let v' = VControl { cvs = state;
+        (* TODO: I am unsure of correct environment *)
+        (* print_endline "doing a control instantiation"; *)
+        let (_,env',st',s) = copyin ctrl env st env typ_decl.constructor_params args' in
+        let state = env'
+          |> EvalEnv.get_val_firstlevel
+          |> List.rev in
+        let v' = VControl { cscope = env;
+                            cvs = state;
                             cparams = typ_decl.params;
                             clocals = typ_decl.locals;
                             apply = typ_decl.apply; } in
         (EvalEnv.pop_scope env',st',s,v')
       | Parser typ_decl ->
-        let (env',st',s) = copyin ctrl env st typ_decl.constructor_params args' in
-        let state = env' |> EvalEnv.get_val_firstlevel |> List.rev in
-        let v' = VParser { pvs = state;
+        (* print_endline "doing a parser instantiation"; *)
+        (* TODO: I am unsure of correct environment *)
+        let (_,env',st',s) = copyin ctrl env st env typ_decl.constructor_params args' in
+        let state = env'
+          |> EvalEnv.get_val_firstlevel
+          |> List.rev in
+        let v' = VParser {pscope = env; 
+                          pvs = state;
                           pparams = typ_decl.params;
                           plocals = typ_decl.locals;
                           states = typ_decl.states; } in
         (EvalEnv.pop_scope env',st',s,v')
       | PackageType pack_decl ->
-        let (env',st',s) = copyin ctrl env st pack_decl.params args' in
-        let state = env' |> EvalEnv.get_val_firstlevel |> List.rev in
+        (* print_endline "doing a package instantiation"; *)
+        (* TODO: I am unsure of correct environment *)
+        let (_,env',st',s) = copyin ctrl env st env pack_decl.params args' in
+        let state = env'
+          |> EvalEnv.get_val_firstlevel
+          |> List.rev in
         (EvalEnv.pop_scope env', st', s, VPackage{decl;args=state})
-      | ExternObject ext_decl -> 
+      | ExternObject ext_decl ->
         let loc = EvalEnv.get_namespace env in
         if State.is_initialized loc st
         then env, st, SContinue, VRuntime {loc = loc; obj_name = (snd ext_decl.name);}
-        else 
+        else
           let args' = List.map args ~f:(fun x -> Some x) in
-          eval_extern_call ctrl env st (snd ext_decl.name) (Some (loc, snd ext_decl.name)) [] args'  
+          (* TODO: I am not sure what the correct lexical scope is... *)
+          eval_extern_call ctrl env st (snd ext_decl.name) (Some (loc, snd ext_decl.name)) [] args'
```

Comment:
Why is `loc` the namespace name? I can't figure out what this code is doing

Reply:
we are using the control-plane names as the locations of externs in the state. the code just grabs the control-plane name out of the environment, where we are storing it, to use in the extern call